pkg/kubelet/kubelet_pods.go

```diff
@@ -1171,6 +1188,32 @@ func (kl *Kubelet) HandlePodCleanups() error {
 	}
 
 	kl.backOff.GC()
+
+	// If two pods with the same UID are observed in rapid succession, we need to
+	// resynchronize the pod worker after the first pod completes and decide whether
+	// to restart the pod. This happens last to avoid confusing the desired state
+	// in other components and to increase the likelihood transient OS failures during
+	// container start are mitigated. In general only static pods will ever reuse UIDs
+	// since the apiserver uses randomly generated UUIDv4 UIDs with a very low
+	// probability of collision.
+	for uid := range restartablePods {
+		pod, ok := allPodsByUID[uid]
+		if !ok {
+			continue
+		}
+		if kl.isAdmittedPodTerminal(pod) {
+			klog.V(3).InfoS("Pod is restartable after termination due to UID reuse, but pod phase is terminal", "pod", klog.KObj(pod), "podUID", pod.UID)
+			continue
+		}
+		start := kl.clock.Now()
+		mirrorPod, _ := kl.podManager.GetMirrorPodByPod(pod)
+		klog.V(3).InfoS("Pod is restartable after termination due to UID reuse", "pod", klog.KObj(pod), "podUID", pod.UID)
+		kl.dispatchWork(pod, kubetypes.SyncPodCreate, mirrorPod, start)
```

Comment:
this basically recreates the HandlePodAdditions work, can we have the shared piece call to a common internal helper?

Reply:
spoke with @smarterclayton about the alternative option of having HandlePodAdditions send back a conflict response when it gets an Add event for a pod that exists under management with the same uid.  There are other complications around that flow for how to best handle requeue that I don't want to block this PR around versus getting back to a functional state.  I think this is my only nit that would like to address so that `HandlePodAdditions` and this area delegate to a common code path that couples dispatchWork and probes together.

Reply:
I would say that the pod worker owns "what is the kubelet doing" and HandlePodAdditions is delivering incremental changes to the pod worker, so I was going to try to minimize what HandlePodAddition does (for better or worse, pod housekeeping is the resync between desired state and actual state).  I *think* that any code that would need to be shared by handle pod addition (which controls admission) and housekeeping (which controls resync of the various loops) should instead live in pod worker, since pod worker owns pod lifecycle.  The probe manager was the only code today, and I wanted to move probe manager addition under pod worker, not outside.  Having a shared function there would actually confuse things a bit more - because I want there to be no shared "make the stuff aware of the start of a pod" (that's pod worker's job)

Reply:
fair enough.